// @electron-forge/maker-squirrel/dist/MakerSquirrel.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("source-map-support/register");

var _makerBase = _interopRequireDefault(require("@electron-forge/maker-base"));

var _electronWinstaller = require("electron-winstaller");

var _fsExtra = _interopRequireDefault(require("fs-extra"));

var _path = _interopRequireDefault(require("path"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class MakerSquirrel extends _makerBase.default {
  constructor(...args) {
    super(...args);

    _defineProperty(this, "name", 'squirrel');

    _defineProperty(this, "defaultPlatforms", ['win32']);
  }

  isSupportedOnCurrentPlatform() {
    return this.isInstalled('electron-winstaller') && !process.env.DISABLE_SQUIRREL_TEST;
  }

  make({
    dir,
    makeDir,
    targetArch,
    packageJSON,
    appName
  }) {
    var _this = this;

    return _asyncToGenerator(function* () {
      const outPath = _path.default.resolve(makeDir, `squirrel.windows/${targetArch}`);

      yield _this.ensureDirectory(outPath);
      const winstallerConfig = Object.assign({
        name: packageJSON.name,
        title: appName,
        noMsi: true,
        exe: `${appName}.exe`,
        setupExe: `${appName}-${packageJSON.version} Setup.exe`
      }, _this.config, {
        appDirectory: dir,
        outputDirectory: outPath
      });
      yield (0, _electronWinstaller.createWindowsInstaller)(winstallerConfig);
      const artifacts = [_path.default.resolve(outPath, 'RELEASES'), _path.default.resolve(outPath, winstallerConfig.setupExe || `${appName}Setup.exe`), _path.default.resolve(outPath, `${winstallerConfig.name}-${packageJSON.version}-full.nupkg`)];

      const deltaPath = _path.default.resolve(outPath, `${winstallerConfig.name}-${packageJSON.version}-delta.nupkg`);

      if (winstallerConfig.remoteReleases || (yield _fsExtra.default.pathExists(deltaPath))) {
        artifacts.push(deltaPath);
      }

      const msiPath = _path.default.resolve(outPath, winstallerConfig.setupMsi || `${appName}Setup.msi`);

      if (!winstallerConfig.noMsi && (yield _fsExtra.default.pathExists(msiPath))) {
        artifacts.push(msiPath);
      }

      return artifacts;
    })();
  }

}

exports.default = MakerSquirrel;
