// @electron-forge/core/dist/util/install-dependencies.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.DepVersionRestriction = exports.DepType = void 0;

require("source-map-support/register");

var _debug = _interopRequireDefault(require("debug"));

var _yarnOrNpm = require("./yarn-or-npm");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

const d = (0, _debug.default)('electron-forge:dependency-installer');
let DepType;
exports.DepType = DepType;

(function (DepType) {
  DepType["PROD"] = "PROD";
  DepType["DEV"] = "DEV";
})(DepType || (exports.DepType = DepType = {}));

let DepVersionRestriction;
exports.DepVersionRestriction = DepVersionRestriction;

(function (DepVersionRestriction) {
  DepVersionRestriction["EXACT"] = "EXACT";
  DepVersionRestriction["RANGE"] = "RANGE";
})(DepVersionRestriction || (exports.DepVersionRestriction = DepVersionRestriction = {}));

var _default =
/*#__PURE__*/
function () {
  var _ref = _asyncToGenerator(function* (dir, deps, depType = DepType.PROD, versionRestriction = DepVersionRestriction.RANGE) {
    d('installing', JSON.stringify(deps), 'in:', dir, `depType=${depType},versionRestriction=${versionRestriction},withYarn=${(0, _yarnOrNpm.hasYarn)()}`);

    if (deps.length === 0) {
      d('nothing to install, stopping immediately');
      return Promise.resolve();
    }

    let cmd = ['install'].concat(deps);

    if ((0, _yarnOrNpm.hasYarn)()) {
      cmd = ['add'].concat(deps);
      if (depType === DepType.DEV) cmd.push('--dev');
      if (versionRestriction === DepVersionRestriction.EXACT) cmd.push('--exact');
    } else {
      if (versionRestriction === DepVersionRestriction.EXACT) cmd.push('--save-exact');
      if (depType === DepType.DEV) cmd.push('--save-dev');
      if (depType === DepType.PROD) cmd.push('--save');
    }

    d('executing', JSON.stringify(cmd), 'in:', dir);

    try {
      yield (0, _yarnOrNpm.yarnOrNpmSpawn)(cmd, {
        cwd: dir,
        stdio: 'pipe'
      });
    } catch (err) {
      throw new Error(`Failed to install modules: ${JSON.stringify(deps)}\n\nWith output: ${err.message}\n${err.stderr.toString()}`);
    }
  });

  return function (_x, _x2) {
    return _ref.apply(this, arguments);
  };
}();

exports.default = _default;
