// @electron-forge/maker-rpm/dist/MakerRpm.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.rpmArch = rpmArch;
exports.default = void 0;

require("source-map-support/register");

var _makerBase = _interopRequireDefault(require("@electron-forge/maker-base"));

var _path = _interopRequireDefault(require("path"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function rpmArch(nodeArch) {
  switch (nodeArch) {
    case 'ia32':
      return 'i386';

    case 'x64':
      return 'x86_64';

    case 'armv7l':
      return 'armv7hl';

    case 'arm':
      return 'armv6hl';

    default:
      return nodeArch;
  }
}

class MakerRpm extends _makerBase.default {
  constructor(...args) {
    super(...args);

    _defineProperty(this, "name", 'rpm');

    _defineProperty(this, "defaultPlatforms", ['linux']);
  }

  isSupportedOnCurrentPlatform() {
    return this.isInstalled('electron-installer-redhat') && process.platform === 'linux';
  }

  make({
    dir,
    makeDir,
    targetArch,
    packageJSON
  }) {
    var _this = this;

    return _asyncToGenerator(function* () {
      const installer = require('electron-installer-redhat');

      const arch = rpmArch(targetArch);
      const name = (_this.config.options || {}).name || packageJSON.name;
      const versionedName = `${name}-${packageJSON.version}.${arch}`;

      const outPath = _path.default.resolve(makeDir, `${versionedName}.rpm`);

      yield _this.ensureFile(outPath);
      const rpmConfig = Object.assign({}, _this.config, {
        arch,
        src: dir,
        dest: _path.default.dirname(outPath),
        rename: undefined
      });
      yield installer(rpmConfig);
      return [outPath];
    })();
  }

}

exports.default = MakerRpm;
